### PR TITLE
aruco refactoring and class interface

### PR DESCRIPTION
**Pull request by AleksandrPanov**
Preparing ArUco module to move to [main opencv](https://github.com/opencv/opencv) repository.
- add `ArucoDetector` class (with detectMarkers() and refineDetectedMarkers() methods)
- class `ArucoDetector` inherits from class `cv::Algorithm` and implemented read()/write()
- add C wrapper for old API (in aruco.hpp/aruco.cpp)
- add aruco_calib_pose.hpp/aruco_calib_pose.cpp and aruco_utils.cpp/aruco_utils.hpp
- add Java and Python tests

_possible add class BaseBoard to work with calibrateCamera/PoseEstimation/SolvePnP?_

**To move ArUco module** to [main opencv](https://github.com/opencv/opencv) repository need to move all files except "aruco.cpp" and "aruco.hpp"

Before merging, it is necessary to make a **git squash**.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/aruco/src/apriltag/apriltag_quad_thresh.cpp [hunk 1]
```diff
@@ -1567,4 +1567,121 @@ imwrite("2.5 debug_lines.pnm", out);
...
 }
 
+void _apriltag(Mat im_orig, const Ptr<DetectorParameters> & _params, std::vector< std::vector< Point2f > > &candidates,
+        std::vector< std::vector< Point > > &contours){
+
+    ///////////////////////////////////////////////////////////
+    /// Step 1. Detect quads according to requested image decimation
+    /// and blurring parameters.
+    Mat quad_im;
+    im_orig.copyTo(quad_im);
```
Comment (asmorkalov) [inline modules/aruco/src/apriltag/apriltag_quad_thresh.cpp:1577]: Most probably it should be in else branch of the next condition to not do extra memory copy for the resize case.

File: modules/aruco/include/opencv2/aruco/board.hpp [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    CV_PROP_RW std::vector< int > ids;
+
+    /// coordinate of the bottom right corner of the board, is set when calling the function create()
+    CV_PROP Point3f rightBottomBorder;
+};
+
+/**
+ * @brief Implementation of drawPlanarBoard that accepts a raw Board pointer.
+ */
+void _drawPlanarBoardImpl(Board *board, Size outSize, OutputArray img, int marginSize = 0, int borderBits = 1);
```
Comment (asmorkalov) [inline modules/aruco/include/opencv2/aruco/board.hpp:75]: It should not be in public header.

Comment (asmorkalov) [inline modules/aruco/include/opencv2/aruco/board.hpp:75]: It should not be in public header.

File: modules/aruco/include/opencv2/aruco_detector.hpp [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+ *   to (I_w_r, I_h_r) = (tau_c / tau_dot_i) * (I_w, I_h), with tau_dot_i = tau_c + max(I_w,I_h) * tau_i.
+ *   Here tau_i (parameter: minMarkerLengthRatioOriginalImg) is a ratio in the range [0,1].
+ *   If we set this to 0, the smallest marker we can detect
+ *   has a side length of tau_c. If we set it to 1 the marker would fill the entire image.
+ *   For a FullHD video a good value to start with is 0.1.
+ * - minSideLengthCanonicalImg: minimum side length of a marker in the canonical image.
+ *   Latter is the binarized image in which contours are searched.
+ *   So all contours with a size smaller than minSideLengthCanonicalImg*minSideLengthCanonicalImg will omitted from the search.
+ * - minMarkerLengthRatioOriginalImg:  range [0,1], eq (2) from paper
+ *   The parameter tau_i has a direct influence on the processing speed.
```
Comment (asmorkalov) [inline modules/aruco/include/opencv2/aruco_detector.hpp:100]: It makes sense to split the description and add description tp each parameter independently.

Comment (asmorkalov) [inline modules/aruco/include/opencv2/aruco_detector.hpp:100]: It makes sense to split the description and add description tp each parameter independently.

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 2 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp:75]: fixed

File: modules/aruco/src/apriltag/apriltag_quad_thresh.cpp (hunk 1 above)
Comment (AleksandrPanov) [reply modules/aruco/src/apriltag/apriltag_quad_thresh.cpp]: fixed

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 2 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp]: fixed

File: modules/aruco/include/opencv2/aruco_detector.hpp (hunk 3 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco_detector.hpp:86]: fixed

Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco_detector.hpp:86]: fixed

File: modules/aruco/include/opencv2/aruco/board.hpp [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    /// The identifiers refers to the board dictionary
+    CV_PROP_RW std::vector< int > ids;
+
+    /// coordinate of the bottom right corner of the board, is set when calling the function create()
+    CV_PROP Point3f rightBottomBorder;
+};
+
+/**
+ * @brief Draw a planar board
+ * @sa _drawPlanarBoardImpl
```
Comment (asmorkalov) [inline modules/aruco/include/opencv2/aruco/board.hpp:74]: Old reference.

File: modules/aruco/src/aruco_detector.cpp [hunk 5]
```diff
@@ -0,0 +1,1230 @@
...
+    // change some parameters to be sure to turn it off
+    if (!params->useAruco3Detection) {
+        params->minMarkerLengthRatioOriginalImg = 0.0;
+        params->minSideLengthCanonicalImg = 0;
+    }
+    else {
+        // always turn on corner refinement in case of Aruco3, due to upsampling
+        params->cornerRefinementMethod = CORNER_REFINE_SUBPIX;
+        // only CORNER_REFINE_SUBPIX implement correctly for useAruco3Detection
+        // TODO: update other CORNER_REFINE methods
```
Comment (asmorkalov) [inline modules/aruco/src/aruco_detector.cpp:809]: Todo

File: modules/aruco/src/aruco_detector.cpp [hunk 6]
```diff
@@ -0,0 +1,1230 @@
...
+                for (int i = range.start; i < range.end; i++) {
+                    _refineCandidateLines(contours[i], candidates[i]);
+                }
+            });
+        }
+    }
+
+    if (params->cornerRefinementMethod != CORNER_REFINE_SUBPIX && fxfy != 1.f) {
+        // only CORNER_REFINE_SUBPIX implement correctly for useAruco3Detection
+        // TODO: update other CORNER_REFINE methods
```
Comment (asmorkalov) [inline modules/aruco/src/aruco_detector.cpp:902]: Todo

File: modules/aruco/src/aruco_utils.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        }
+    }
+    else {
+        CV_Error(cv::Error::StsNotImplemented,
+                 "Only Mat vector, UMat vector, and vector<vector> OutputArrays are currently supported.");
+    }
+}
+
+/*
+std::vector<std::vector<Point2f> > getVectors(InputArrayOfArrays& in) {
```
Comment (asmorkalov) [inline modules/aruco/src/aruco_utils.cpp:42]: Please remove dead code.

File: modules/aruco/include/opencv2/aruco/board.hpp [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+ * @param outSize size of the output image in pixels.
+ * @param img output image with the board. The size of this image will be outSize
+ * and the board will be on the center, keeping the board proportions.
+ * @param marginSize minimum margins (in pixels) of the board in the output image
+ * @param borderBits width of the marker borders.
+ *
+ * This function return the image of a planar board, ready to be printed. It assumes
+ * the Board layout specified is planar by ignoring the z coordinates of the object points.
+ */
+CV_EXPORTS_W void drawPlanarBoard(const Ptr<Board> &board, Size outSize, OutputArray img,
```
Comment (asmorkalov) [inline modules/aruco/include/opencv2/aruco/board.hpp:87]: Looks like `drawPlanarBoard` and the next `GridBoard::draw` and `CharucoBoard::draw` have the same semantic. I propose to add virtual method to base class and overload it for all children to have unified interface.

**Comment by asmorkalov** [comment]
cc @alalek @vpisarev @victor1234 @savuor

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 4 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp]: fixed

File: modules/aruco/src/aruco_detector.cpp (hunk 5 above)
Comment (AleksandrPanov) [reply modules/aruco/src/aruco_detector.cpp]: fixed

File: modules/aruco/src/aruco_detector.cpp (hunk 6 above)
Comment (AleksandrPanov) [reply modules/aruco/src/aruco_detector.cpp]: fixed

File: modules/aruco/src/aruco_utils.cpp (hunk 7 above)
Comment (AleksandrPanov) [reply modules/aruco/src/aruco_utils.cpp]: fixed

**Comment by asmorkalov** [comment]
Thanks a lot for the patch! I tested the solution manually and it works for me. Couple of general comments on the code:
- It make sense to duplicate some of detector tests, but with old API and stay it in contrib after migration. It allows to ensure that wrappers are in "calable" state and presumes original behaviour. They should be in separate test file for easy extraction.
- There are still several "todo" in code like "// Todo: update other CORNER_REFINE methods". Do you plan to implement this after merge?

File: modules/aruco/src/apriltag/apriltag_quad_thresh.hpp [hunk 9]
```diff
@@ -121,5 +122,15 @@ void threshold(const Mat mIm, const Ptr<DetectorParameters> &parameters, Mat& mT
  */
 zarray_t *apriltag_quad_thresh(const Ptr<DetectorParameters> &parameters, const Mat & mImg, std::vector< std::vector< Point > > &contours);
 
+/**
+ *
+ * @param im_orig
+ * @param _params
+ * @param candidates
+ * @param contours
+ */
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/src/apriltag/apriltag_quad_thresh.hpp]: fixed, removed empty docstring for private function

File: modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+enum PatternPos {
+    /** @brief The marker coordinate system is centered on the middle of the marker.
+        * The coordinates of the four corners (CCW order) of the marker in its own coordinate system are:
+        * (-markerLength/2, markerLength/2, 0), (markerLength/2, markerLength/2, 0),
+        * (markerLength/2, -markerLength/2, 0), (-markerLength/2, -markerLength/2, 0).
+        *
+        * These pattern points define this coordinate system:
+        * ![Image with axes drawn](images/singlemarkersaxes.jpg)
+        */
+    CCW_center,
```
Comment (vpisarev) [inline modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp:31]: if I remember correctly our discussion, shall we add some description on which coordinate system is future-proof and compatible with other detectors and which is obsolete?

File: modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+ * (markerLength/2, -markerLength/2, 0), (-markerLength/2, -markerLength/2, 0)
+ * @sa use cv::drawFrameAxes to get world coordinate system axis for object points
+ * @sa @ref tutorial_aruco_detection
+ * @sa EstimateParameters
+ * @sa PatternPos
+ */
+CV_EXPORTS_W void estimatePoseSingleMarkers(InputArrayOfArrays corners, float markerLength,
+                                            InputArray cameraMatrix, InputArray distCoeffs,
+                                            OutputArray rvecs, OutputArray tvecs, OutputArray _objPoints = noArray(),
+                                            Ptr<EstimateParameters> estimateParameters = EstimateParameters::create());
```
Comment (vpisarev) [inline modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp:105]: for now it's fine, but when we move these functions to the main opencv module, can we consider using named parameters concept, which will provide more convenient API, especially for Python and modern C++ users? (https://github.com/opencv/opencv/wiki/OE-34.-Named-Parameters)

File: modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+ * receives a list of detected corners and its identifiers from several views of the Board.
+ * The function returns the final re-projection error.
+ */
+CV_EXPORTS_AS(calibrateCameraCharucoExtended)
+double calibrateCameraCharuco(InputArrayOfArrays charucoCorners, InputArrayOfArrays charucoIds,
+                              const Ptr<CharucoBoard> &board, Size imageSize, InputOutputArray cameraMatrix,
+                              InputOutputArray distCoeffs, OutputArrayOfArrays rvecs, OutputArrayOfArrays tvecs,
+                              OutputArray stdDeviationsIntrinsics, OutputArray stdDeviationsExtrinsics,
+                              OutputArray perViewErrors, int flags = 0, TermCriteria criteria = TermCriteria(
+                                      TermCriteria::COUNT + TermCriteria::EPS, 30, DBL_EPSILON));
```
Comment (vpisarev) [inline modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp:269]: same question about extended vs non-extended

File: modules/aruco/include/opencv2/aruco/board.hpp [hunk 13]
```diff
@@ -0,0 +1,229 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+#ifndef __OPENCV_BOARD_HPP__
```
Comment (vpisarev) [inline modules/aruco/include/opencv2/aruco/board.hpp:4]: __OPENCV_ARUCO_BOARD_HPP__?

File: modules/aruco/include/opencv2/aruco/charuco.hpp [hunk 14]
```diff
@@ -42,6 +42,8 @@ the use of this software, even if advised of the possibility of such damage.
 #include <opencv2/core.hpp>
```
Comment (vpisarev) [inline modules/aruco/include/opencv2/aruco/charuco.hpp:42]: many changes -> can we put the new shorter license header?

**Comment by asmorkalov** [comment]
@vpisarev The PR presume old API for existing code and introduce the new one that will be migrated to main repo. That's why original and extended functions are presumed. The same for the coordinate system orientation. OpenCV uses rig corner as zero point. Aruco original code uses rig center as zero point. Some options are introduced to presume compatibility.

File: modules/aruco/include/opencv2/aruco_detector.hpp [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+enum CornerRefineMethod{
+    CORNER_REFINE_NONE,     ///< Tag and corners detection based on the ArUco approach
+    CORNER_REFINE_SUBPIX,   ///< ArUco approach and refine the corners locations using corner subpixel accuracy
+    CORNER_REFINE_CONTOUR,  ///< ArUco approach and refine the corners locations using the contour-points line fitting
+    CORNER_REFINE_APRILTAG, ///< Tag and corners detection based on the AprilTag 2 approach @cite wang2016iros
+};
+
+struct CV_EXPORTS_W DetectorParameters {
+    DetectorParameters() {
+        adaptiveThreshWinSizeMin = 3;
```
Comment (asmorkalov) [inline modules/aruco/include/opencv2/aruco_detector.hpp:53]: The structure fields are not documented.

File: modules/aruco/src/aruco_detector.cpp (hunk 5 above)
Comment (asmorkalov) [reply modules/aruco/src/aruco_detector.cpp]: There are several todos for CORNER_REFINE_SUBPIX in the same file. Please take a look.

File: modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp (hunk 10 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp:31]: fixed

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 13 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp]: fixed

File: modules/aruco/include/opencv2/aruco/charuco.hpp (hunk 14 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/charuco.hpp:7]: fixed

added shorter license to all modified files

File: modules/aruco/include/opencv2/aruco_detector.hpp [hunk 16]
```diff
@@ -0,0 +1,408 @@
...
+//! @{
+
+enum CornerRefineMethod{
+    CORNER_REFINE_NONE,     ///< Tag and corners detection based on the ArUco approach
+    CORNER_REFINE_SUBPIX,   ///< ArUco approach and refine the corners locations using corner subpixel accuracy
+    CORNER_REFINE_CONTOUR,  ///< ArUco approach and refine the corners locations using the contour-points line fitting
+    CORNER_REFINE_APRILTAG, ///< Tag and corners detection based on the AprilTag 2 approach @cite wang2016iros
+};
+
+struct CV_EXPORTS_W DetectorParameters {
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco_detector.hpp:56]: reference added

File: modules/aruco/include/opencv2/aruco_detector.hpp [hunk 17]
```diff
@@ -0,0 +1,408 @@
...
+ * This module is dedicated to square fiducial markers (also known as Augmented Reality Markers)
+ * These markers are useful for easy, fast and robust camera pose estimation.ç
+ *
+ * The main functionalities are:
+ * - Detection of markers in an image
+ * - Pose estimation from a single marker or from a board/set of markers
+ * - Detection of ChArUco board for high subpixel accuracy
+ * - Camera calibration from both, ArUco boards and ChArUco boards.
+ * - Detection of ChArUco diamond markers
+ * The samples directory includes easy examples of how to use the module.
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco_detector.hpp]: docs updated

File: modules/aruco/include/opencv2/aruco_detector.hpp [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+    /// dictionary indicates the type of markers that will be searched
+    CV_PROP_RW Ptr<Dictionary> dictionary;
+
+    /// marker detection parameters
+    CV_PROP_RW Ptr<DetectorParameters> params;
+
+    /// marker refine parameters
+    CV_PROP_RW Ptr<RefineParameters> refineParams;
+
+    ArucoDetector(const Ptr<Dictionary> &_dictionary = getPredefinedDictionary(DICT_4X4_50), const Ptr<DetectorParameters> &_params = DetectorParameters::create(),
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco_detector.hpp]: docs updated

**Review by alalek — COMMENTED** [review]
Several notes about public API part are below.

File: modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+enum PatternPos {
+    /** @brief The marker coordinate system is centered on the middle of the marker.
+     * The coordinates of the four corners (CCW order) of the marker in its own coordinate system are:
+     * (-markerLength/2, markerLength/2, 0), (markerLength/2, markerLength/2, 0),
+     * (markerLength/2, -markerLength/2, 0), (-markerLength/2, -markerLength/2, 0).
+     *
+     * These pattern points define this coordinate system:
+     * ![Image with axes drawn](images/singlemarkersaxes.jpg)
+     */
+    CCW_CENTER,
```
Comment (alalek) [inline modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp:30]: Non-class enum values should be properly prefixed to avoid conflicts.

File: modules/aruco/include/opencv2/aruco/board.hpp [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+
+    private:
+    // number of markers in X and Y directions
+    int _markersX, _markersY;
+
+    // marker side length (normally in meters)
+    float _markerLength;
+
+    // separation between markers in the grid
+    float _markerSeparation;
```
[the remark this answers is not in the document]
Comment (alalek) [reply modules/aruco/include/opencv2/aruco/board.hpp:142]: no changes and no objectives here

File: modules/aruco/include/opencv2/aruco/board.hpp [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+
+/**
+ * @brief test whether the ChArUco markers are collinear
+ *
+ * @param board layout of ChArUco board.
+ * @param charucoIds list of identifiers for each corner in charucoCorners per frame.
+ * @return bool value, 1 (true) if detected corners form a line, 0 (false) if they do not.
+      solvePnP, calibration functions will fail if the corners are collinear (true).
+ *
+ * The number of ids in charucoIDs should be <= the number of chessboard corners in the board.  This functions checks whether the charuco corners are on a straight line (returns true, if so), or not (false).  Axis parallel, as well as diagonal and other straight lines detected.  Degenerate cases: for number of charucoIDs <= 2, the function returns true.
```
Comment (alalek) [inline modules/aruco/include/opencv2/aruco/board.hpp:220]: It is better to split very long lines. Keep one sentence per line (useful to track modifications too).

File: modules/aruco/include/opencv2/aruco_detector.hpp [hunk 22]
```diff
@@ -0,0 +1,441 @@
...
+    CV_PROP_RW bool useAruco3Detection;
+
+    /// minimum side length of a marker in the canonical image. Latter is the binarized image in which contours are searched.
+    CV_PROP_RW int minSideLengthCanonicalImg;
+
+    /// range [0,1], eq (2) from paper. The parameter tau_i has a direct influence on the processing speed.
+    CV_PROP_RW float minMarkerLengthRatioOriginalImg;
+
+private:
+    bool readWrite(const Ptr<FileNode>& readNode = nullptr, const Ptr<FileStorage>& writeStorage = nullptr);
```
Comment (alalek) [inline modules/aruco/include/opencv2/aruco_detector.hpp:240]: Why do we need this in public OpenCV API?

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp:138]: Should I use the `pimple` interface like in the QR detector? Then I'll hide the fields in an opaque index.

File: modules/aruco/include/opencv2/aruco/board.hpp [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+
+    /// the dictionary of markers employed for this board
+    CV_PROP Ptr<Dictionary> dictionary;
+
+    /// vector of the identifiers of the markers in the board (same size than objPoints)
+    /// The identifiers refers to the board dictionary
+    CV_PROP_RW std::vector<int> ids;
+
+    /// coordinate of the bottom right corner of the board, is set when calling the function create()
+    CV_PROP Point3f rightBottomBorder;
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp]: Then the fields need to be hidden?

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp:138]: Or do I need to use inheritance like in the `AccumLayer`/`AccumLayerImpl` class?

File: modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp [hunk 24]
```diff
@@ -0,0 +1,286 @@
...
+        * These pattern points define this coordinate system:
+        * ![Image with axes drawn](images/singlemarkersaxes2.jpg)
+        *
+        * These pattern dots are convenient to use with a chessboard/ChArUco board.
+        */
+    CW_top_left_corner
+};
+
+/** @brief
+ * Pose estimation parameters
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp]: fixed

File: modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp [hunk 25]
```diff
@@ -0,0 +1,286 @@
...
+ * (markerLength/2, -markerLength/2, 0), (-markerLength/2, -markerLength/2, 0)
+ * @sa use cv::drawFrameAxes to get world coordinate system axis for object points
+ * @sa @ref tutorial_aruco_detection
+ * @sa EstimateParameters
+ * @sa PatternPos
+ */
+CV_EXPORTS_W void estimatePoseSingleMarkers(InputArrayOfArrays corners, float markerLength,
+                                            InputArray cameraMatrix, InputArray distCoeffs,
+                                            OutputArray rvecs, OutputArray tvecs, OutputArray _objPoints = noArray(),
+                                            Ptr<EstimateParameters> estimateParameters = EstimateParameters::create());
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp]: fixed

File: modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp [hunk 26]
```diff
@@ -0,0 +1,286 @@
...
+ * (-markerLength/2, markerLength/2, 0), (markerLength/2, markerLength/2, 0),
+ * (markerLength/2, -markerLength/2, 0), (-markerLength/2, -markerLength/2, 0)
+ * @sa use cv::drawFrameAxes to get world coordinate system axis for object points
+ * @sa @ref tutorial_aruco_detection
+ * @sa EstimateParameters
+ * @sa PatternPos
+ */
+CV_EXPORTS_W void estimatePoseSingleMarkers(InputArrayOfArrays corners, float markerLength,
+                                            InputArray cameraMatrix, InputArray distCoeffs,
+                                            OutputArray rvecs, OutputArray tvecs, OutputArray _objPoints = noArray(),
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp]: fixed to just `objPoints`

File: modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp [hunk 27]
```diff
@@ -0,0 +1,286 @@
...
+ * This function calibrates a camera using an Aruco Board. The function receives a list of
+ * detected markers from several views of the Board. The process is similar to the chessboard
+ * calibration in calibrateCamera(). The function returns the final re-projection error.
+ */
+CV_EXPORTS_AS(calibrateCameraArucoExtended)
+double calibrateCameraAruco(InputArrayOfArrays corners, InputArray ids, InputArray counter, const Ptr<Board> &board,
+                            Size imageSize, InputOutputArray cameraMatrix, InputOutputArray distCoeffs,
+                            OutputArrayOfArrays rvecs, OutputArrayOfArrays tvecs, OutputArray stdDeviationsIntrinsics,
+                            OutputArray stdDeviationsExtrinsics, OutputArray perViewErrors, int flags = 0,
+                            TermCriteria criteria = TermCriteria(TermCriteria::COUNT + TermCriteria::EPS, 30, DBL_EPSILON));
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp]: fixed

File: modules/aruco/include/opencv2/aruco/board.hpp [hunk 28]
```diff
@@ -0,0 +1,229 @@
...
+ * A board is a set of markers in the 3D space with a common coordinate system.
+ * The common form of a board of marker is a planar (2D) board, however any 3D layout can be used.
+ * A Board object is composed by:
+ * - The object points of the marker corners, i.e. their coordinates respect to the board system.
+ * - The dictionary which indicates the type of markers of the board
+ * - The identifier of all the markers in the board.
+ */
+class CV_EXPORTS_W Board {
+    public:
+   /**
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp]: fixed indentation

File: modules/aruco/include/opencv2/aruco/board.hpp [hunk 29]
```diff
@@ -0,0 +1,229 @@
...
+    * @brief Provide way to create Board by passing necessary data. Specially needed in Python.
+    *
+    * @param objPoints array of object points of all the marker corners in the board
+    * @param dictionary the dictionary of markers employed for this board
+    * @param ids vector of the identifiers of the markers in the board
+    *
+    */
+    CV_WRAP static Ptr<Board> create(InputArrayOfArrays objPoints, const Ptr<Dictionary> &dictionary, InputArray ids);
+
+   /**
```
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp]: fixed

File: modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp (hunk 19 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/aruco_calib_pose.hpp]: add `ARUCO_` prefix

File: modules/aruco/include/opencv2/aruco_detector.hpp (hunk 22 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco_detector.hpp:240]: removed

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 21 above)
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp]: splitted

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp:138]: Maybe you shouldn't create an Impl class for a single class?

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp]: These fields are used by the custom code. Should I break backward compatibility?

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp]: @alalek hid the field to`GridBoardImpl/CharucoBoardImpl` and added pointer to `GridBoardImpl/CharucoBoardImpl`

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco/board.hpp]: @alalek
Potential solutions:
- just add `getter`/`setter`, move fields to private in 5.x
- rename the `class` to `struct `and save direct access to fields (This "Board" data structure is quite simple.)

**Comment by AleksandrPanov** [comment]
@alalek, `gridImpl `and `charucoImpl` were added. Need help with getters and setters.

File: modules/aruco/include/opencv2/aruco/board.hpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (alalek) [reply modules/aruco/include/opencv2/aruco/board.hpp]: As there is a plan to provide aruco API through the main opencv repository, then we should propose the robust API for that.

Direct access properties, especially with "write" access are dangerous. E.g., user could "shoot the leg" writing wrong `ids` values. We should avoid providing such API which allows to make object into inconsistent state.

For backward compatibility we could provide compatibility code between legacy and new API.

**Review by vpisarev — APPROVED** [review]

File: modules/aruco/include/opencv2/aruco_detector.hpp (hunk 22 above)
Comment (asmorkalov) [reply modules/aruco/include/opencv2/aruco_detector.hpp:240]: @AleksandrPanov It's still there.

Comment (AleksandrPanov) [reply modules/aruco/include/opencv2/aruco_detector.hpp]: removed `readWrite()` from public API

**Review by asmorkalov — APPROVED** [review]
:+1: LGTM

**Comment by alalek** [comment]
> `@asmorkalov merged commit 8eaa8ac into opencv:4.x`

All 11 commits are merged. They should be squashed instead.

File: modules/aruco/src/aruco.cpp [hunk 30]
```diff
@@ -1,1904 +1,32 @@
...
-disclaimed. In no event shall copyright holders or contributors be liable for
-any direct, indirect, incidental, special, exemplary, or consequential damages
-(including, but not limited to, procurement of substitute goods or services;
-loss of use, data, or profits; or business interruption) however caused
-and on any theory of liability, whether in contract, strict liability,
-or tort (including negligence or otherwise) arising in any way out of
-the use of this software, even if advised of the possibility of such damage.
-*/
-
-#include "precomp.hpp"
```
Comment (alalek) [inline modules/aruco/src/aruco.cpp:39]: `precomp.hpp` must be restored.

File: modules/aruco/test/test_precomp.hpp [hunk 31]
```diff
@@ -7,7 +7,6 @@
 #include "opencv2/ts.hpp"
 #include "opencv2/imgproc.hpp"
 #include "opencv2/calib3d.hpp"
-#include "opencv2/aruco.hpp"
 #include <opencv2/aruco/charuco.hpp>
```
Comment (alalek) [inline modules/aruco/test/test_precomp.hpp:10]: What is the purpose of this change?

If you want to remove module's headers, then remove them all (charuco.hpp is still here).

Currently it looks like that something was broken and there is a hack to hide that (so user code may be broken too).
